Jest tests cause require-dir to break in dependencies

<!-- 
THIS IS NOT A HELP FORUM. 
If you are experiencing problems with setting up Jest, please make sure to visit our Help page: 
https://facebook.github.io/jest/help.html
-->
<!-- 
Before creating an issue please check the following:
* you are using the latest version of Jest
* try re-installing your node_modules folder 
* run Jest once with `--no-cache` to see if that fixes the problem you are experiencing. 
-->

**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
When a Jest test requires a module in `node_modules` that uses [`require-dir`](https://github.com/aseemk/requireDir) with a relative path, the path is resolved relatively to the package that has the executed Jest test instead of the package that uses `require-dir` (what's supposed to happen). This issue does not occur if exactly the same code is executed directly through Node instead of Jest.

**If the current behavior is a bug, please provide the steps to reproduce and either a repl.it demo through https://repl.it/languages/jest or a minimal repository on GitHub that we can `yarn install` and `yarn test`.**

1. Clone https://github.com/nickmccurdy/jest-require-dir-bug
2. `npm install-test`

```
~/Repos/jest-require-dir-bug → node index.test.js 
~/Repos/jest-require-dir-bug → npm test

> @ test /Users/nick/Repos/ncu-example
> jest

 FAIL  ./index.test.js
  ● Test suite failed to run

    ENOENT: no such file or directory, scandir '/Users/nick/Repos/ncu-example/dir'
      
      at Object.fs.readdirSync (fs.js:911:18)
      at requireDir (node_modules/require-dir/index.js:25:20)
      at Object.<anonymous> (node_modules/example/index.js:1:112)
      at Object.<anonymous> (index.test.js:1:90)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.88s
Ran all test suites.
npm ERR! Test failed.  See above for more details.
```

This issue can also be observed by requiring code that uses `npm-check-updates` (which also uses `require-dir`) via JavaScript in a Jest test.

**What is the expected behavior?**
Jest's execution environment should allow `require-dir` to function normally, requiring a relative directory with the `requireDir('./example')` syntax even in a dependency. Jest should also behave the same as Node itself, or work around this difference in execution environments.

**Please provide your exact Jest configuration and mention your Jest, node, yarn/npm version and operating system.**
- Empty configuration
- `index.test.js` or `__tests__/index.js` naming conventions
- Jest 15-21.2.1
- `require-dir` 0.3.2
- Node 8.6.0
- npm 5.4.2
- macOS 10.13

--- a/packages/jest-cli/src/__tests__/search_source.test.js
+++ b/packages/jest-cli/src/__tests__/search_source.test.js
@@ -373,12 +373,14 @@ describe('SearchSource', () => {
 
     it('finds tests that depend directly on the path', () => {
       const filePath = path.join(rootDir, 'RegularModule.js');
+      const file2Path = path.join(rootDir, 'RequireRegularModule.js');
       const loggingDep = path.join(rootDir, 'logging.js');
       const parentDep = path.join(rootDir, 'ModuleWithSideEffects.js');
       const data = searchSource.findRelatedTests(new Set([filePath]));
       expect(toPaths(data.tests).sort()).toEqual([
         parentDep,
         filePath,
+        file2Path,
         loggingDep,
         rootPath,
       ]);

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -38,7 +38,6 @@ describe('Runtime requireModule', () => {
         'filename',
         'id',
         'children',
-        'parent',
         'paths',
         'require',
       ]);

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -50,7 +50,7 @@ describe('Runtime requireModule', () => {
         runtime.__mockRootPath,
         'RegularModule',
       );
-      expect(exports.parent).toBeUndefined();
+      expect(exports.parent).toBeNull();
     }));
 
   it('resolve module.parent.require correctly', () =>

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -36,7 +36,17 @@ describe('Runtime requireModule', () => {
         exports: {},
         filename: 'mock.js',
         id: 'mockParent',
+        require: expect.any(Function),
       });
+    }));
+
+  it('resolve module.parent.require correctly', () =>
+    createRuntime(__filename).then(runtime => {
+      const exports = runtime.requireModule(
+        runtime.__mockRootPath,
+        'inner_parent_module',
+      );
+      expect(exports).toEqual('This should happen');
     }));
 
   it('provides `module.filename` to modules', () =>

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -38,9 +38,9 @@ describe('Runtime requireModule', () => {
         'filename',
         'id',
         'children',
+        'parent',
         'paths',
         'require',
-        'parent',
       ]);
     }));
 

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -31,14 +31,26 @@ describe('Runtime requireModule', () => {
     createRuntime(__filename).then(runtime => {
       const exports = runtime.requireModule(
         runtime.__mockRootPath,
+        'RequireRegularModule',
+      );
+      expect(Object.keys(exports.parent)).toEqual([
+        'exports',
+        'filename',
+        'id',
+        'children',
+        'paths',
+        'require',
+        'parent',
+      ]);
+    }));
+
+  it('`module.parent` should be undefined for entrypoints', () =>
+    createRuntime(__filename).then(runtime => {
+      const exports = runtime.requireModule(
+        runtime.__mockRootPath,
         'RegularModule',
       );
-      expect(exports.parent).toEqual({
-        exports: {},
-        filename: '',
-        id: '',
-        require: expect.any(Function),
-      });
+      expect(exports.parent).toBeUndefined();
     }));
 
   it('resolve module.parent.require correctly', () =>
@@ -58,7 +70,7 @@ describe('Runtime requireModule', () => {
       );
 
       expect(slash(exports.parentFileName.replace(__dirname, ''))).toEqual(
-        '/test_root/node_modules/module-needing-parent/index.js',
+        '/test_root/inner_parent_module.js',
       );
     }));
 

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -8,7 +8,8 @@
 
 'use strict';
 
-const path = require('path');
+import path from 'path';
+import slash from 'slash';
 
 let createRuntime;
 
@@ -34,7 +35,7 @@ describe('Runtime requireModule', () => {
       );
       expect(exports.parent).toEqual({
         exports: {},
-        filename: 'mock.js',
+        filename: '',
         id: 'mockParent',
         require: expect.any(Function),
       });
@@ -46,7 +47,19 @@ describe('Runtime requireModule', () => {
         runtime.__mockRootPath,
         'inner_parent_module',
       );
-      expect(exports).toEqual('This should happen');
+      expect(exports.outputString).toEqual('This should happen');
+    }));
+
+  it('resolve module.parent.filename correctly', () =>
+    createRuntime(__filename).then(runtime => {
+      const exports = runtime.requireModule(
+        runtime.__mockRootPath,
+        'inner_parent_module',
+      );
+
+      expect(slash(exports.parentFileName.replace(__dirname, ''))).toEqual(
+        '/test_root/node_modules/module-needing-parent/index.js',
+      );
     }));
 
   it('provides `module.filename` to modules', () =>

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -36,7 +36,7 @@ describe('Runtime requireModule', () => {
       expect(exports.parent).toEqual({
         exports: {},
         filename: '',
-        id: 'mockParent',
+        id: '',
         require: expect.any(Function),
       });
     }));

--- a/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
+++ b/packages/jest-runtime/src/__tests__/runtime_require_module.test.js
@@ -38,6 +38,7 @@ describe('Runtime requireModule', () => {
         'filename',
         'id',
         'children',
+        'parent',
         'paths',
         'require',
       ]);

--- a/packages/jest-runtime/src/__tests__/test_root/RequireRegularModule.js
+++ b/packages/jest-runtime/src/__tests__/test_root/RequireRegularModule.js
@@ -1,0 +1,12 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @providesModule RequireRegularModule
+ */
+
+'use strict';
+
+module.exports.parent = require('RegularModule').parent;

--- a/packages/jest-runtime/src/__tests__/test_root/inner_parent_module.js
+++ b/packages/jest-runtime/src/__tests__/test_root/inner_parent_module.js
@@ -1,0 +1,14 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @providesModule inner_parent_module
+ */
+
+'use strict';
+
+const impl = require('module-needing-parent');
+
+module.exports = impl;

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -492,19 +492,21 @@ class Runtime {
 
     const dirname = path.dirname(filename);
     localModule.children = [];
-    localModule.paths = this._resolver.getModulePaths(dirname);
-    localModule.require = this._createRequireImplementation(filename, options);
 
     Object.defineProperty(
       localModule,
       'parent',
       // https://github.com/facebook/flow/issues/285#issuecomment-270810619
       ({
+        enumerable: true,
         get() {
           return moduleRegistry[from];
         },
       }: Object),
     );
+
+    localModule.paths = this._resolver.getModulePaths(dirname);
+    localModule.require = this._createRequireImplementation(filename, options);
 
     const transformedFile = this._scriptTransformer.transform(
       filename,

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -77,7 +77,6 @@ const getModuleNameMapper = (config: ProjectConfig) => {
 
 const mockParentModule = {
   exports: {},
-  filename: 'mock.js',
   id: 'mockParent',
 };
 
@@ -497,6 +496,7 @@ class Runtime {
     localModule.paths = this._resolver.getModulePaths(dirname);
     localModule.require = this._createRequireImplementation(filename, options);
     localModule.parent = Object.assign({}, localModule.parent, {
+      filename: lastExecutingModulePath,
       require: this._createRequireImplementation(
         lastExecutingModulePath,
         options,

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -496,6 +496,12 @@ class Runtime {
     localModule.parent = mockParentModule;
     localModule.paths = this._resolver.getModulePaths(dirname);
     localModule.require = this._createRequireImplementation(filename, options);
+    localModule.parent = Object.assign({}, localModule.parent, {
+      require: this._createRequireImplementation(
+        lastExecutingModulePath,
+        options,
+      ),
+    });
 
     const transformedFile = this._scriptTransformer.transform(
       filename,

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -323,7 +323,7 @@ class Runtime {
         // $FlowFixMe
         localModule.exports = require(modulePath);
       } else {
-        this._execModule(localModule, options);
+        this._execModule(localModule, options, moduleRegistry[from]);
       }
     }
     return moduleRegistry[modulePath].exports;
@@ -472,7 +472,11 @@ class Runtime {
     return to ? this._resolver.resolveModule(from, to) : from;
   }
 
-  _execModule(localModule: Module, options: ?InternalModuleOptions) {
+  _execModule(
+    localModule: Module,
+    options: ?InternalModuleOptions,
+    parent?: Module,
+  ) {
     // If the environment was disposed, prevent this module from being executed.
     if (!this._environment.global) {
       return;
@@ -489,15 +493,7 @@ class Runtime {
     localModule.children = [];
     localModule.paths = this._resolver.getModulePaths(dirname);
     localModule.require = this._createRequireImplementation(filename, options);
-    localModule.parent = {
-      exports: {},
-      filename: lastExecutingModulePath,
-      id: lastExecutingModulePath,
-      require: this._createRequireImplementation(
-        lastExecutingModulePath,
-        options,
-      ),
-    };
+    localModule.parent = parent;
 
     const transformedFile = this._scriptTransformer.transform(
       filename,

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -57,6 +57,8 @@ type CoverageOptions = {
   mapCoverage: boolean,
 };
 
+type ModuleRegistry = {[key: string]: Module};
+
 type BooleanObject = {[key: string]: boolean};
 type CacheFS = {[path: Path]: string};
 
@@ -86,13 +88,13 @@ class Runtime {
   _currentlyExecutingModulePath: string;
   _environment: Environment;
   _explicitShouldMock: BooleanObject;
-  _internalModuleRegistry: {[key: string]: Module};
+  _internalModuleRegistry: ModuleRegistry;
   _isCurrentlyExecutingManualMock: ?string;
   _mockFactories: {[key: string]: () => any};
   _mockMetaDataCache: {[key: string]: MockFunctionMetadata};
   _mockRegistry: {[key: string]: any};
   _moduleMocker: ModuleMocker;
-  _moduleRegistry: {[key: string]: Module};
+  _moduleRegistry: ModuleRegistry;
   _resolver: Resolver;
   _shouldAutoMock: boolean;
   _shouldMockModuleCache: BooleanObject;
@@ -475,7 +477,7 @@ class Runtime {
   _execModule(
     localModule: Module,
     options: ?InternalModuleOptions,
-    moduleRegistry: {[key: string]: Module},
+    moduleRegistry: ModuleRegistry,
     from: Path,
   ) {
     // If the environment was disposed, prevent this module from being executed.

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -502,7 +502,7 @@ class Runtime {
       ({
         enumerable: true,
         get() {
-          return moduleRegistry[from];
+          return moduleRegistry[from] || null;
         },
       }: Object),
     );

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -475,7 +475,7 @@ class Runtime {
   _execModule(
     localModule: Module,
     options: ?InternalModuleOptions,
-    parent?: Module,
+    parentModule?: Module,
   ) {
     // If the environment was disposed, prevent this module from being executed.
     if (!this._environment.global) {
@@ -491,9 +491,9 @@ class Runtime {
 
     const dirname = path.dirname(filename);
     localModule.children = [];
+    localModule.parent = parentModule;
     localModule.paths = this._resolver.getModulePaths(dirname);
     localModule.require = this._createRequireImplementation(filename, options);
-    localModule.parent = parent;
 
     const transformedFile = this._scriptTransformer.transform(
       filename,

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -323,7 +323,7 @@ class Runtime {
         // $FlowFixMe
         localModule.exports = require(modulePath);
       } else {
-        this._execModule(localModule, options, moduleRegistry[from]);
+        this._execModule(localModule, options, moduleRegistry, from);
       }
     }
     return moduleRegistry[modulePath].exports;
@@ -384,7 +384,7 @@ class Runtime {
         filename: modulePath,
         id: modulePath,
       };
-      this._execModule(localModule);
+      this._execModule(localModule, undefined, this._mockRegistry, from);
       this._mockRegistry[moduleID] = localModule.exports;
     } else {
       // Look for a real module to generate an automock from
@@ -475,7 +475,8 @@ class Runtime {
   _execModule(
     localModule: Module,
     options: ?InternalModuleOptions,
-    parentModule?: Module,
+    moduleRegistry: {[key: string]: Module},
+    from: Path,
   ) {
     // If the environment was disposed, prevent this module from being executed.
     if (!this._environment.global) {
@@ -491,9 +492,19 @@ class Runtime {
 
     const dirname = path.dirname(filename);
     localModule.children = [];
-    localModule.parent = parentModule;
     localModule.paths = this._resolver.getModulePaths(dirname);
     localModule.require = this._createRequireImplementation(filename, options);
+
+    Object.defineProperty(
+      localModule,
+      'parent',
+      // https://github.com/facebook/flow/issues/285#issuecomment-270810619
+      ({
+        get() {
+          return moduleRegistry[from];
+        },
+      }: Object),
+    );
 
     const transformedFile = this._scriptTransformer.transform(
       filename,

--- a/packages/jest-runtime/src/index.js
+++ b/packages/jest-runtime/src/index.js
@@ -75,11 +75,6 @@ const getModuleNameMapper = (config: ProjectConfig) => {
   return null;
 };
 
-const mockParentModule = {
-  exports: {},
-  id: 'mockParent',
-};
-
 const unmockRegExpCache = new WeakMap();
 
 class Runtime {
@@ -492,16 +487,17 @@ class Runtime {
 
     const dirname = path.dirname(filename);
     localModule.children = [];
-    localModule.parent = mockParentModule;
     localModule.paths = this._resolver.getModulePaths(dirname);
     localModule.require = this._createRequireImplementation(filename, options);
-    localModule.parent = Object.assign({}, localModule.parent, {
+    localModule.parent = {
+      exports: {},
       filename: lastExecutingModulePath,
+      id: lastExecutingModulePath,
       require: this._createRequireImplementation(
         lastExecutingModulePath,
         options,
       ),
-    });
+    };
 
     const transformedFile = this._scriptTransformer.transform(
       filename,

